WearOAuth/oauth-device-grant/src/main/java/com/example/android/wearable/oauth/devicegrant/AuthDeviceGrantViewModel.kt

```diff
@@ -0,0 +1,183 @@
+ * device, then polls for the access token, and uses it to retrieve the user's name.
+ */
+class AuthDeviceGrantViewModel(private val remoteIntentHelper: RemoteIntentHelper) : ViewModel() {
+    // Status to show on the Wear OS display
+    val status: MutableLiveData<String> by lazy { MutableLiveData<String>() }
+    private fun showStatus(statusString: String) = status.postValue(statusString)
+
+    fun startAuthFlow() {
+        viewModelScope.launch(Dispatchers.IO) {
+            // Step 1: Retrieve the verification URI
+            showStatus("Starting authorization... Switch to your phone to authenticate.")
+            val verificationInfo = retrieveVerificationInfo().getOrElse {
+                showStatus("Authorization failed")
+                return@launch
+            }
+
+            // Step 2: Show the pairing code & open the verification URI on the paired device
+            showStatus("code: ${verificationInfo.userCode}")
+            fireRemoteIntent(verificationInfo.verificationUri)
+
+            // Step 3: Poll the Auth server for the token
+            val token = retrieveToken(verificationInfo.deviceCode, verificationInfo.interval)
+
+            // Step 4: Use the token to make an authorized request
+            val userName = retrieveUserProfile(token).getOrElse {
+                showStatus("Authorization failed")
+                return@launch
+            }
+
+            showStatus("User profile retrieved. Welcome $userName!")
+        }
+    }
+
+    // The response data when retrieving the verification
+    data class VerificationInfo(
+        val verificationUri: String,
+        val userCode: String,
+        val deviceCode: String,
+        val interval: Int
+    )
+
+    /**
+     * Retrieve the information needed to verify the user. When performing this request, the server
+     * generates a user & device code pair. The user code is shown to the user and opened on the
+     * paired device. The device code is passed while polling the OAuth server.
+     */
+    private suspend fun retrieveVerificationInfo(): Result<VerificationInfo> {
+        return try {
+            Log.d(TAG, "Retrieving verification info...")
+            val responseJson = doPostRequest(
+                url = "https://oauth2.googleapis.com/device/code",
+                params = mapOf(
+                    "client_id" to CLIENT_ID,
+                    "scope" to "https://www.googleapis.com/auth/userinfo.profile"
+                )
+            )
+            Result.success(
+                VerificationInfo(
+                    verificationUri = responseJson.getString("verification_url"),
+                    userCode = responseJson.getString("user_code"),
+                    deviceCode = responseJson.getString("device_code"),
+                    interval = responseJson.getInt("interval")
+                )
+            )
+        } catch (e: Exception) {
```

Comment:
This `Exception` will catch `CancellationException` in addition to network errors. We probably want to catch the cancellation exceptions separately and rethrow, so that we don't convert them into a `Result.failure`.

Reply:
I added another catch for all the try/catch blocks. Nice catch! (pun inteded)